### ISSUE TITLE
Summary functions and multiple exchanges for same metabolite

### ISSUE THREAD

**Issue by matthiaskoenig** [opencobra/cobrapy#441]
Hi all,

somehow cobrapy has problems if there are reactions which are similar to exchange reactions.
In the example below the FBA solution is completely unbalanced
EX_Glcxt': -1000.0 and 'EX_O2': -1000.0 !
Glcxt and O2 definitely not balanced.

There are exchange reactions on the metabolites with correct direction and SBO terms in the SBML. Somehow the alternative reactions are just dropped from the optimization. There seems to be a bug if there are two reactions which change a single metabolite (SBML & network image attached).

```
import cobra
from pprint import pprint

source = 'diauxic_fba.xml'
mfba = cobra.io.read_sbml_model(source)
s = mfba.optimize(objective_sense="maximize")

print(s)
print(mfba.solution.status)

print
pprint(mfba.solution.x_dict)

# summary of solution
print
mfba.summary(fva=True)
```

```
<Solution 92.44 at 0x7f271ffa5190>
optimal

{'EX_Ac': 168.1666610317663,
 'EX_Glcxt': -1000.0,
 'EX_O2': -1000.0,
 'EX_X': 92.43550856585347,
 'v1': 0.0,
 'v2': 78.56037151702786,
 'v3': 0.0,
 'v4': 13.875137048825605,
 'vGlcxt': 10.0,
 'vO2': 15.0}

IN FLUXES                OUT FLUXES                  OBJECTIVES
-----------------------  --------------------------  ------------
id       Flux  Range     id      Flux  Range         v1   0
-----  ------  --------  ----  ------  ------------  v2  78.6
O2         15  [15, 15]  Ac     168    [168, 168]    v3   0
Glcxt      10  [10, 10]  X       92.4  [92.4, 92.4]  v4  13.9
```

[diauxic_fba.xml.zip](https://github.com/opencobra/cobrapy/files/822487/diauxic_fba.xml.zip)
![network](https://cloud.githubusercontent.com/assets/900538/23629549/575e13c0-02b8-11e7-9fec-26e6f664a84f.png)

**Comment by cdiener** [comment]
Could you define the your expected behavior. Is the problem the discrepancy between the summary and you import fluxes?

**Comment by matthiaskoenig** [comment]
I screwed up with the stoichiometric coefficients. Solution is balanced. Sorry for that.

Only thing remaining is that model.summary is not displaying all exchange fluxes for O2 and Glcxt. I.e. if there are multiple boundary reactions, all should be displayed in the summary.

```
IN FLUXES                OUT FLUXES                  OBJECTIVES
-----------------------  --------------------------  ------------
id       Flux  Range     id      Flux  Range         v1   0
-----  ------  --------  ----  ------  ------------  v2  78.6
O2         1000  [1000, 1000]  Ac     168    [168, 168]    v3   0
Glcxt      1000  [1000, 1000]  X       92.4  [92.4, 92.4]  v4  13.9
O2         15  [15, 15]
Glcxt      10  [10, 10]
```

**Comment by cdiener** [comment]
Glad it worked out :smile: Yes, I think summary currently assumes that a metabolite has at most one exchange reaction. Pinging @pstjohn who knows more about that.

**Comment by matthiaskoenig** [comment]
The system_boundary reactions are correct. 
```
 	lb 	ub 	reversibility 	objective 	boundary
EX_O2 	-1000 	1000 	True 	NaN 	system_boundary
EX_Glcxt 	-1000 	1000 	True 	NaN 	system_boundary
EX_Ac 	-1000 	1000 	True 	NaN 	system_boundary
EX_X 	-1000 	1000 	True 	NaN 	system_boundary
vO2 	0 	15 	False 	NaN 	system_boundary
vGlcxt 	0 	10 	False 	NaN 	system_boundary
v1 	0 	1000 	False 	1 	None
v2 	0 	1000 	False 	1 	None
v3 	0 	1000 	False 	1 	None
v4 	0 	1000 	False 	1 	None
```


The bug is here in the `model.model_summary` function (code below).
You overwrite the metabolite_fluxes if there are multiple boundary reactions for a metabolite. You iterate over the `boundary_reactions`, but store via `metabolite_fluxes[met].update`. This does not work.

```
 for rxn in boundary_reactions:
        for met, stoich in iteritems(rxn.metabolites):
            metabolite_fluxes[met] = {
                'id': format_long_string(met.id, 15),
                'flux': stoich * rxn.x}

            if fva:
                imin = stoich * fva_results.loc[rxn.id]['minimum']
                imax = stoich * fva_results.loc[rxn.id]['maximum']

                # Correct 'max' and 'min' for negative values
                metabolite_fluxes[met].update({
                    'fmin': imin if abs(imin) <= abs(imax) else imax,
                    'fmax': imax if abs(imin) <= abs(imax) else imin,
                })
```

**Comment by matthiaskoenig** [comment]
Just to clear things up. 
The `devel-2 branch` is the one against which I should do a pull request? Also this is the branch where the new solver things are developed (from cameo)?

Is there any documentation about what is best practice to commit a bug fix.

**Comment by hredestig** [comment]
Yes, devel-2 is the current main branch but just for a few more hours :) You can make your pr against devel-2, if needed I'll switch it for you later. Yup, there are [guidelines for contributing](https://github.com/opencobra/cobrapy/blob/devel/CONTRIBUTING.rst)

**Comment by pstjohn** [comment]
So this might be a fundamental flaw in reporting the model summary by metabolite name rather than reaction..

For instance, in this situation

O2 <=>    rxn1, bounds = (-1000, 1000)
O2 <=>    rxn2, bounds = (-1000, 1000)
O2 --> my_objective    rxn3, bounds = (-1000, 1000)

FVA would tell you that independently, rxn1 and rxn2 have bounds of (-1000, 1000), but you'd be incorrect if you said the bound on O2 import was the sum (-2000, 2000). We'd need FVA on (rxn1 + rxn2), which isn't really supported by the current methods, correct?

Also, [these lines](https://github.com/opencobra/cobrapy/blob/devel-2/cobra/flux_analysis/summary.py#L142-L146) would also cause multiple imports to screw up.

So I would recommend we either:
1. recast `model_summary` in terms of reactions to support this type of model construction, or
2. raise a warning when we detect that there are multiple exchanges for each metabolite.

**Comment by matthiaskoenig** [comment]
You can just report the sum of fluxes for the metabolite.
Same for FVA: You can just run the FVA and report the max and min sum of all exchange reactions for a metabolite than you have the min/max of FVA for this metabolite exchange.
Than you have just one reported exchange flux which can be the sum of multiple actual exchange fluxes for the metabolites.
In addition show a warning in this cases so that the user is aware.

**Comment by pstjohn** [comment]
> Same for FVA: You can just run the FVA and report the max and min sum of all exchange reactions for a metabolite than you have the min/max of FVA for this metabolite exchange.

My point above is that it's not that simple. For many metabolites, the limiting factor on import is not the bounds on the specific exchange reactions, but the capacity of the model to use that metabolite. If you only look at FVA for reactions individually, you lose that information. (Above, the maximum ability of the model to use O2 is 1000, but summing the FVA results would yield 2000).

**Comment by matthiaskoenig** [comment]
I understand. I thought is is easily possible to calculate this.

One would mainly have to write a custom version of
calculate_lp_variability
which looks for the maximum flux through the sum of all the exchange reactions for a metabolite (just set the objective coefficient to 1 (-1) for all exchange reactions (depending on import/export direction).

Write now for me it is a good solution to provide the sum of all exchange reactions for a metabolite (i.e. the net exchange) and a warning that the FVA may be wrong due to multiple exchange reactions.

**Comment by cdiener** [comment]
Agree with @pstjohn for FVA. I think the best option for now is to report on the level of reactions only.

**Comment by sumanasri** [comment]
HI,

Is there a way to put a constaint on the sum of influxes or outfluxes to or from a metabolite? suppose X is a metabolite and I1, I2 and I3 are fluxes that produce it. Can I say something like I1+I2+I3 upper bound = 7?

**Comment by pstjohn** [comment]
Also, was this issue closed by #718?

**Comment by Midnighter** [comment]
Thought so by the title but hadn't read the whole thread until now.
